keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `Local` policy for `service.spec.externalTrafficPolicy` and
+`service.spec.internalTrafficPolicy` allows configuring services so
+that they will always deliver to local endpoints, avoiding an
+unnecessary hop to another node when the request could just as easily
+be processed locally.
+
+However, they both fail entirely if there are no endpoints for the
+service on the node. This tradeoff is unacceptable for some use cases.
+
+The [Topology Aware Hints feature] provides a way of signaling to
+kube-proxy that it should optimize service delivery on a
+per-availability zone basis, but does not provide any way to optimize
+on any other basis.
+
+This KEP adds a new topology hint, to tell kube-proxy that a Service
+is expected to have an endpoint on every node most of the time, and so
+it should optimize traffic by always routing to a local endpoint _when
+one is available_, but using other topology hints or ordinary
+`Cluster` traffic policy when not.
+
+[Topology Aware Hints feature]: ../2433-topology-aware-hints/
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
```

Comment:
Will there be any mechanism to prevent a local endpoint from getting overloaded? With zonal hints that was already a significant risk, but with a single node the risk seems to increase exponentially. If there isn't a good way to mitigate this risk, I think this would need to be opt-in until we had some kind of feedback loop that enabled us to fallover to other endpoints when the local one had reached capacity.

Reply:
The design is really all based on the DNS user story (since that's the only user story we currently have).

In the case of a CoreDNS pod on every node, the assumption is that no endpoint will ever get overloaded; no one ever does that much DNS (unless they're doing some sort of DoS attack in which case all bets are off). Although kube-proxy is _normally_ responsible for trying to prevent endpoints from getting overloaded, we are saying that for _this_ kind of service, it does not have that responsibility.

I guess I can add some hedging to the description to allow for the possibility that in the future, node-level topology might send traffic to other nodes even when there is an endpoint on this node, if it thinks that that would result in a faster reply. But I didn't have any intention of trying to implement that behavior now.

Reply:
The main difference I see between "policy" and "topology" is that policy is a user-provided expression of intent and "topology" is a collection of heuristics that we apply (eventually without the user saying anything).  If we're ever going to auto-enable this heuristic, how can we build in more safety?  If this is not going to be auto-enabled ever, is it topology or policy?

Reply:
It feels like there must be more use cases for a PreferLocal approach here, but I don't know what those are. If we're focusing on DNS, it seems like something that is going to be configured once per cluster and therefore a perfect candidate for opt-in functionality and not trying to automatically detect this scenario.

I think a good dividing line here is that topology has tried to include some form of logic to determine if it's safe to enable. We're trying to guess what would be a reasonable approach for a user. If we believe that topology in Kubernetes will ever have a feedback loop that allows us to understand when an endpoint has reached capacity, same-node routing would also be in scope. Unfortunately that seems to require _very significant_ architectural changes. 

What is proposed here does not include anything like that. Instead it is simply "route to local endpoints if they exist, otherwise revert to default behavior". That seems like it's closer in nature to xTP=Local and it seems to fully solve the DNS use case. This all leads me to think that it _may_ be worth having the following options available:

* xTP=Local
* iTP=PreferLocal

Since xTP doesn't have any kind of fallback mechanism for local traffic, we're really just interested in what would happen for internal traffic. My theory is that it would be reasonable for iTP to fallback to topology hints if it was enabled, or "cluster" if not.

To summarize, I think it's probably simplest to consider this policy and not try to build any advanced logic into this. That means we'd be agreeing that:

* The simple implementation of routing to local endpoints if present is sufficient for the use cases we are aware of
* This is not something we'll ever want to enable by default
* This will generally only be enabled on one Service per cluster
* It is unlikely that we'll have a feedback loop in Kubernetes that allows us to safely spillover when a local endpoint reaches capacity

Reply:
One more thought here, I've gotten at least one feature request to just have an equivalent to the "PreferNode" concept here but for zone. Essentially "if any endpoints are in my local zone, route to them". Based on the criteria above, that also does not seem to fit into topology aware hints but would be more similar to whatever this "PreferNode" concept becomes, if we want to support it. In both cases, the users are saying they can ensure that endpoints are distributed appropriately, they just want a way to ensure traffic stays within the bounds they want.

Reply:
FTR: I was previously against "PreferLocal" as a policy, but the discussion has softened me on that.